File: 0912-sort-an-array/0912-sort-an-array.java
class Solution {
    public int[] sortArray(int[] arr) {
        // Arrays.sort(nums);
        // return nums;
        
        //using heap sort - Time : O(nlogn) , Space : O(1)
        int n = arr.length;
        for(int i = n / 2 - 1; i>=0; i--){
            heapify(arr,n,i);
        }
        
        for(int i = n - 1; i>0; i--){
            int temp = arr[0];
            arr[0] = arr[i];
            arr[i] = temp;
            
            heapify(arr,i,0);
        }
        return arr;
    }
    
    public void heapify(int []arr , int n , int i){
        int largest = i;
        int left = i * 2 + 1;
        int right = i * 2 + 2;
        
        if(left < n && arr[left] > arr[largest])largest = left;
        if(right < n && arr[right] > arr[largest])largest = right;
        
        if(largest != i){
            int temp = arr[largest];
            arr[largest] = arr[i];
            arr[i] = temp;
            
            //recursively calling heap sort
            heapify(arr , n , largest);
        }
    }
}

File: 0912-sort-an-array/0912-sort-an-array.py
class Solution:
    def sortArray(self, nums: List[int]) -> List[int]:
        
        n = len(nums)
        for i in range (n//2 -1,-1,-1):
            self.heapify(nums,n,i)
            
        for i in range(n-1,0,-1):
            temp = nums[i];
            nums[i] = nums[0]
            nums[0] = temp;
            
            self.heapify(nums , i , 0)
        return nums
    
    def heapify(self , nums : List[int] , n : int , i : int):
        largest = i
        left = i * 2 + 1
        right = i * 2 + 2
        
        if left < n and nums[left] > nums[largest]:
            largest = left
        if right < n and nums[right] > nums[largest]:
            largest = right
            
        if largest != i :
            nums[i] , nums[largest] = nums[largest] , nums[i]
            
            self.heapify(nums , n , largest)    
        
            
        